### ISSUE TITLE
Enums support

### ISSUE THREAD

**Issue by Egorand** [realm/realm-java#776]
I was hoping enums are supported, as they're a better alternative to ints for enumerations.

**Comment by cmelchior** [comment]
Hi @Egorand 
Enums are on our wishlist as well, but there are some challenges involved as one of the primary design goals of Realm is cross-platform compatibility and enums doesn't exist in Objective-C like they do in Java.

**Comment by Egorand** [comment]
Hi @cmelchior and thanks for the response,
Enums can easily be converted into simple ints and back, I think typedefs are the closest feature that Objective-C provides (my knowledge of Objective-C is close to zero). Hence it shouldn't be hard to represent enums internally in a way that both platforms could work with it.

**Comment by cmelchior** [comment]
True, but that would be very brittle with regard to modifying the enums afterwards. See here for a discussion on the topic: http://stackoverflow.com/questions/229856/ways-to-save-enums-in-database.

It is possible to find a solution, but right now we want to prioritise other features.

**Comment by mpost** [comment]
I would also like to see enum support.

**Comment by hamen** [comment]
+1

**Comment by frazer-rbsn** [comment]
+1!

**Comment by ganlaw** [comment]
+1

**Comment by rgrinberg** [comment]
+1

**Comment by Saketme** [comment]
+1

**Comment by MrZeb** [comment]
+1

**Comment by AKlimashevskyCedon** [comment]
is it possible to help you to speed up enum support?

**Comment by StErMi** [comment]
Any updates?

**Comment by cmelchior** [comment]
Right now we are focusing on other bigger features (custom methods, null, async queries and migration API). However once we open up for custom methods it will be possible to work around this by doing the conversion yourself in the getter/setter. 

That said we still want to add support for it, but proper support will take a bit longer.

**Comment by benjamincombes** [comment]
+1

**Comment by benjamincombes** [comment]
I don't know if it can help anybody, but here is what I use as a workaround :
```java
public enum CampaignStatus {
    LIVE,
    UPCOMING
}

@SerializedName("status")
private String statusRaw;
@Ignore
private transient CampaignStatus status;

public CampaignStatus getStatus() {
    return CampaignStatus.valueOf(statusRaw.toUpperCase());
}

public void setStatus(CampaignStatus status) {
    statusRaw = status.name();
}
```

**Comment by mateusgrb** [comment]
+1

**Comment by paulpv** [comment]
+1

**Comment by musesum** [comment]
+1

**Comment by doncorsean** [comment]
+1

**Comment by mehrad-rafigh** [comment]
+1

**Comment by ddszczygiel** [comment]
+1

**Comment by Queatz** [comment]
+1

**Comment by cmelchior** [comment]
Note that with #2196 now merged to master (will be part of 0.88),  it is possible to support Enums using the following pattern:

```
public enum MyEnum {
  FOO, BAR;
}

public class Foo extends RealmObject {
  private String enumDescription;

  public void saveEnum(MyEnum val) {
    this.enumDescription = val.toString();
  }

  public MyEnum getEnum() {
    return MyEnum.valueOf(enumDescription):
  }
}
```

**Comment by proverbface** [comment]
+1

**Comment by njovy** [comment]
+1

**Comment by PaulWoitaschek** [comment]
@cmelchior 
Use `#Enum.name()` instead of `toString()` as its final.

**Comment by oscarg798** [comment]
+1

**Comment by mateusgrb** [comment]
Guys, you can now react to issues/comments with a thumbs up. No more need to comment "+1" =)

**Comment by Ozrenko** [comment]
What do you think about using annotations instead of enums? This is one of recommendation that comes from google. More info you can get here: https://www.youtube.com/watch?v=Hzs6OBcvNQE

**Comment by JoeRock11** [comment]
+1

**Comment by cruisediary** [comment]
+1

**Comment by n-belokopytov** [comment]
+1

**Comment by ashk3156** [comment]
my realm object class is this
`public enum ProductType {
        ACTIVE, IN_ACTIVE, OUT_OF_STOCK;
    }
public void saveEnum(ProductType val) {

        if (!this.isValid) {
            this.enumvalue = val.IN_ACTIVE.toString().toUpperCase();
        } else if (this.quantity_ <= 0) {
            this.enumvalue = val.OUT_OF_STOCK.toString().toUpperCase();
        } else {
            this.enumvalue = val.ACTIVE.toString().toUpperCase();
        }
    }

    public ProductType getEnum() {
        return ProductType.valueOf(enumvalue);
    }`
but when i am calling getEnum, it is giving null exception.why it is not initialising  ?@cmelchior

**Comment by cmelchior** [comment]
@ashk3156 Please create another issue than reusing this one, and please add the code that actually throw the NullPointer as well.

**Comment by romshiri** [comment]
+1

**Comment by nicomlas** [comment]
+1

**Comment by mateusgrb** [comment]
Reporting a case where lack of enums support made things less pretty: in Kotlin, I created an interface Contact and a class ContactRealm implementing that interface. The idea was to use Contact wherever possible and keep its ~implementation~ decoupled, so that later I could plug a ContactORM implementing same interface, for example. Contact has a field 'sex' which would be best represented as an enum but due to Realm's limitation I had to declare it as a String in the first place, otherwise wouldn't be able to override the field in ContactRealm later.

**Comment by Zhuinden** [comment]
> otherwise wouldn't be able to override the field in ContactRealm later.

?

**Comment by mateusgrb** [comment]
interface Contact {
        var name: String
        var age: Int
        var pictureUrl: String
        var sex: String

        enum class Sex {
            MALE, FEMALE
        }
    }

    open class ContactRealm(
            override var name: String = "",
            override var age: Int = 0,
            override var pictureUrl: String = "",
            override var sex: String = "") : RealmObject(), Contact

@Zhuinden the field `sex`in Contact can't be a `Sex`, as it needs to be overridden in ContactRealm and Realm doesn't support enums.

**Comment by powder366** [comment]
+1

**Comment by mptrista** [comment]
+1

**Comment by apouche** [comment]
I ran into this thread while trying to solve the same problem. Here was my approach to this in kotlin

```
open class Contract: RealmModel {
    enum class PaymentMethod { CREDIT_CARD, WIRE_TRANSFER }
    private var _paymentMethod: String? = null
   var paymentMethod: PaymentMethod?
        get() {
            if (_paymentMethod != null)
                return PaymentMethod.valueOf(_paymentMethod!!)
            else
                return null
        }
        set(value) {
            val realm = Realm.getDefaultInstance()
            realm.executeTransaction {
                _paymentMethod = value.toString()
            }
            realm.close()

        }
}
````

basically I used a private string property taken as a string to construct a computed property with that enum correspondance. Not sure this is very "realm" friendly since I'm new to Android development but I thought this worked actually pretty well.

**Comment by YuriDenison** [comment]
+1

**Comment by Zhuinden** [comment]
I think this is a subset of https://github.com/realm/realm-java/issues/1694 .

**Comment by cmelchior** [comment]
Yes, it will most likely be implemented that way since the concept of a Java enum doesn't easily transfer to other platforms.

**Comment by eygraber** [comment]
I investigated ObjectBox today (don't worry; I'm sticking with Realm 😜 ), and came up with something like [this for getting enums supported](https://github.com/objectbox/objectbox-java/issues/287). Would something like that work here?

**Comment by PerfectedTech** [comment]
+1

**Comment by cfxway** [comment]
+1

**Comment by blocha** [comment]
still no support? after so many years?

**Comment by shreshtha18** [comment]
how to store list of enum in realm android?

**Comment by Zhuinden** [comment]
`RealmList<String>` although then it becomes non-queryable as per https://github.com/realm/realm-object-store/issues/513 so it should probably be combined to a comma separated list in a single String value.

**Comment by shaqaruden** [comment]
How is this still not supported?

**Comment by Ashkan-san** [comment]
2024 and still no real enum support?
